### PR TITLE
Add scalar event logging function

### PR DESCRIPTION
**Pull request by jeffcarp**
Currently there are three `jax.monitoring` methods for recording events:

- `record_event(event: str, **kwargs: str | int)`
- `record_event_duration_secs(event: str, duration: float, **kwargs: str | int)`
- `record_event_time_span(event: str, start_time: float, end_time: float, **kwargs: str | int)`

This PR adds a more general method, `record_scalar`:

- `record_scalar(event: str, value: float | int, **kwargs: str | int)`

which enables JAX core and its ecosystem libraries to emit a wider more flexible set of metrics, for example FLOPS usage, gradient norm/clipping stats, quantization ranges, and more. We (@jshin1394 and I) are also working on a plugin that will make it easy to send `jax.monitoring` metrics to TensorBoard and other tools.

### CODE REVIEW

**Review by froystig — APPROVED** [review]

**Comment by jeffcarp** [comment]
@froystig thank you for the review! Is there a way for me to re-run the CI? I think it failed with an unrelated build error.

**Comment by froystig** [comment]
I re-ran the checks but they failed again. If you think these errors are unrelated and fixed, then you could try to rebase on upstream `main` and then force-push the PR branch. Doing so ought to also trigger another re-run.

**Comment by jeffcarp** [comment]
Thanks, rebased and looks like they're passing now.
